File: algorithms/python/numberofRecentCalls/numberofRecentCalls.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/number-of-recent-calls/
# Author: Miao Zhang
# Date:   2021-03-26

class RecentCounter:

    def __init__(self):
        self.q = collections.deque()

    def ping(self, t: int) -> int:
        while self.q and self.q[0] < t - 3000:
            self.q.popleft();
        self.q.append(t)
        return len(self.q)


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)

File: algorithms/cpp/numberofRecentCalls/numberofRecentCalls.cpp
// Source: https://leetcode.com/problems/number-of-recent-calls/
// Author: Miao Zhang
// Date:   2021-03-26

class RecentCounter {
public:
    RecentCounter() {

    }

    int ping(int t) {
        while (!q.empty() && t - q.front() > 3000) q.pop();
        q.push(t);
        return q.size();
    }

private:
    queue<int> q;
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */
